### PR TITLE
core: Fix bullet append from html (fix #6072)

### LINKED ISSUE #6072: Text bullets should not appear on blank lines
### Describe the bug

Ruffle shows bullet points on blank lines.

image

### Expected behavior

Flash player does not 
image

### Affected platform

Desktop app

### Operating system

Windows 10

### Browser

_No response_

### Additional information

SWF file:
[bullets.zip](https://github.com/ruffle-rs/ruffle/files/7876889/bullets.zip)

### PR DESCRIPTION
**Pull request by FernandoKGA**
Adds a boolean to check either if the line is empty and with this information not appending a bullet at the beginning of the line.
However I've reached a strange solution, where I need to verify if the text line is not empty, and if it's not, the code should not append the bullet to it, it will leave to the next line (very strange behavior) to append, this next line will be a `\n`, so will generate a empty text and then append.

I'm not confident with this fix, if @Herschel could take a look and say some thoughts about it, I would appreciate, as a rookie Rustacean, this fix maybe don't make sense on the project perspective.

Fixes #6072 

TODO:
- [x] Check if the logic is correct
- [ ] Write tests for it

### CODE REVIEW

**Comment by Herschel** [comment]
Thank you! Bear with me as I'm not too familiar myself with how Flash handles HTML text layout.

With this PR, the standard case of a bullet list without newlines is incorrect, as `text.is_empty()` is false in these cases:
[bullet.zip](https://github.com/ruffle-rs/ruffle/files/7921843/bullet.zip)
image

(It seems incorrect on master, too, instead showing an extra bullet at the end:)
image

Instead of passing the `empty_text_line` bool around, I wonder if we can determine it directly in `fixup_line` because we are already iterating through the layout boxes for that line.

File: core/src/html/layout.rs [hunk 1]
```diff
@@ -278,8 +279,7 @@ impl<'a, 'gc> LayoutContext<'a, 'gc> {
         } else {
             self.font_leading_adjustment()
         };
-
-        if self.current_line_span.bullet && self.is_first_line {
+        if self.current_line_span.bullet && self.is_first_line && empty_text_line.unwrap_or(false) {
```
Comment (Herschel) [inline core/src/html/layout.rs:282]: I wonder if instead we can determine that the line is empty immediately here by looking at the `boxes` or `box_count`. Maybe:
`if self.current_line_span.bullet && self.is_first_line && box_count > 0`
Because if there are no boxes, then there is no text; but I'm not sure if this catches all edge cases. Will need testing to see if whitespace affects whether a bullet exists or not, etc.

**Comment by danielhjacobs** [comment]
A good test case for bullets is here: https://dnalc.cshl.edu/view/2022-G2C-3-D-Brain.html

Choose any option from the dropdown and scroll down; it all looks right with Flash Player.

There are also missing line breaks, but that's because h tags are used alongside a stylesheet, and stylesheets are unsupported in Ruffle, so you can ignore that.

**Comment by danielhjacobs** [comment]
https://dnalc.cshl.edu/view/2022-G2C-3-D-Brain.html currently:
associated_functions_current
substructures_current_top
substructures_current_bottom
research_reviews_current_top
research_reviews_current_bottom_and_links

https://dnalc.cshl.edu/view/2022-G2C-3-D-Brain.html with just the changes from this branch:
associated_functions_empty
substructures_empty_bottom
substructures_empty_top
research_reviews_empty_top
research_reviews_empty_bottom_and_links

https://dnalc.cshl.edu/view/2022-G2C-3-D-Brain.html with just the changes from https://github.com/ruffle-rs/ruffle/pull/5758:
associated_functions_htags
substructures_htags_top
substructures_htags_bottom
research_reviews_htags_top
research_reviews_htags_bottom_and_links

https://dnalc.cshl.edu/view/2022-G2C-3-D-Brain.html with the changes from this branch plus the changes from https://github.com/ruffle-rs/ruffle/pull/5758:
associated_functions_both
substructures_both_top
substructures_both_bottom
research_reviews_both_top
research_reviews_both_bottom_and_links


https://dnalc.cshl.edu/view/2022-G2C-3-D-Brain.html with Flash player:
associated_functions_flash
substructures_flash_top
substructures_flash_bottom
research_reviews_flash_top
research_reviews_flash_bottom_and_links

**Comment by danielhjacobs** [comment]
On the other hand, here's Mike suggestion in conjunction with https://github.com/ruffle-rs/ruffle/pull/5758 looks just like official Flash player.

File: core/src/html/layout.rs (hunk 1 above)
Comment (danielhjacobs) [reply core/src/html/layout.rs:282]: Testing a change which just added this line and the changes from https://github.com/ruffle-rs/ruffle/pull/5758, I found that this emulated the bullets on https://dnalc.cshl.edu/view/2022-G2C-3-D-Brain.html perfectly.

**Comment by danielhjacobs** [comment]
Oh, I poorly phrased my previous comments. You should remove the h1-h6 commit. Despite it usually being more accurate, it was not the proper

**Comment by danielhjacobs** [comment]
Oh, I poorly framed previous comments.  You should remove the 2nd commit.  Despite it often making things look more accurate it is not the proper way to do things and was decided against.  I just used it for testing this branch because it made things look better for the test file I was using

**Comment by FernandoKGA** [comment]
Just implemented the new check with `box_count` and seems to be working more accurately as @Herschel said. Now I will look forward on how to test this correctly with as many tests I can (never done it before, probably will take some time). @danielhjacobs If you have any other insights about this PR or how I should proceed, feel free to write down! 😄

**Comment by danielhjacobs** [comment]
I mentioned this on Discord already. I looked into this a bit more:

While the bullets are not displayed, it seems the TextFormat related to them is retained:

I added a trace(text.htmlText) to the file Mike provided, and changed the array to ["Header","<li>bullet 1","","bullet 2","bullet 3","bullet 4</li>"], and using that as a test file, on the branch from this PR the test says:

```
[INFO  ruffle_core::player] Loaded SWF version 6, with a resolution of 550x400
thread 'bullet_empty_lines' panicked at 'assertion failed: `(left == right)`: ruffle output != flash player output

Diff < left / right > :
```

Where the diff is the existence of `<TEXTFORMAT LEADING="2"><LI><FONT FACE="Noto Sans" SIZE="12" COLOR="#000000" LETTERSPACING="0" KERNING="0"></FONT></LI></TEXTFORMAT>` in the location of the empty bullet in Flash, but not in Ruffle.

I think that means the proper change would be even if the line is set to have a bullet, don't display a bullet if it's empty, and also means that the only possible test would be visual.

**Comment by Herschel** [comment]
I don't think a test is necessary here -- these kind of changes are unfortunately hard to test (purely visual, won't match Flash unless we're pixel perfect, etc.) and are bound to change often as we improve our text rendering.

Possibly we could add a general image test for text, similar to the `color` test, that adds a reference rendering for a movie with a bunch of text.

**Comment by danielhjacobs** [comment]
I just posted this on Discord:

"If [this PR] were made not a draft, I think it would be immediately merge-able. [You] wanted to add tests, but as Mike pointed out, it's hard to test a purely visual change like that. ... I wanted to check if any core devs agreed with me that it should probably just be merged as-is."

"I had thought I found an issue with it, but I'm fairly sure I must have done the test wrong, since the issue I thought I found would make no sense to have changed."

Adrian responded: "At a glance, it looks safe... Not immediate idea about how to do a visual test, but other than that it's indeed merge-able"

So I think if you are okay with not adding tests (or perhaps adding visual tests separately in a future PR), it would be fine. if you marked this PR as ready for review.

**Comment by FernandoKGA** [comment]
> 

No problem at all marking it as ready to review. I'm taking some time to do the tests because I try to contribute when I have some free time. If other devs think it's ok, I will do it.

**Review by relrelb — APPROVED** [review]

**Review by adrian17 — APPROVED** [review]

**Comment by relrelb** [comment]
Thank you!
